### PR TITLE
oop-special_class_method-hard-q6

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# oop-special_class_method-hard-q6
+
+## Question Text
+
+You are writing a Java program that mimics different Vehicles and their attributes. Write a Java class called **Car**
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/QuestionText.md:5]: This paragraph has a lot of information made redundant by the specifications below.

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+that has a `double` variable called **topSpeed** and another `String` variable called **type**. For this
+class, write a constructor that accepts no parameters, but sets the variables **topSpeed** and **type** to 180 and null,
+respectively. Also, write a constructor that accepts a `double` and a `String` value, and sets the values
+of the **topSpeed** and **type** variable to those values, respectively. Also, write a **toString** method that returns
+a summary of the attributes of the object. Also, write accessor and mutator methods for the necessary variables.
+
+### Car Class:
+
+Your `Car` class must have the following items:
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/QuestionText.md:14]: change field to attribute

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    - a `String` field named **type** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts a`double` and a `String` to initialize the
+      **topSpeed** and **type** fields.
+
+- Methods:
+    - a `setTopSpeed` method which accepts a `double` value, updates the **topSpeed** field with it, and is accessible
+      from
+      other classes
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/QuestionText.md:27]: fix formatting

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    - a `getTopSpeed` method which is accessible from other classes and returns the value of the **topSpeed** field
+    - a `setType` method which accepts a `String` value, updates the **type** field with it, and is accessible
+      from other classes
+    - a `getType` method which is accessible from other classes and returns the value of the **type** field
+    - a `toString` method which returns a String of the information about the `Car` class of the form
+      `"Car{topSpeed: " + topSpeed + ", type: " + type + "}"`.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/QuestionText.md:33]: change toString to the format agreed upon in the meeting throughout the question

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class MainTest {
+    // Java
+    private final String className = "Car";
+    private final String doubleFieldName = "topSpeed";
+    private final String StringFieldName = "type";
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:23]: change variable name to not start with an uppercase letter

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setStringMethodName, "does not correctly set the value of the", StringFieldName, "field.");
+        Object setMethodOutput = testClass.callMethod(setStringMethodName, setMethodArguments, setMethodModifiers, classInstance);
+        _assertEquals(updatedValue, testClass.getFieldValue(classInstance, StringFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setStringMethodName, "should not return any output"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void chairClassHasCorrectGetTypeMethod(double value1, String value2) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:193]:
```suggestion
    public void carClassHasCorrectGetTypeMethod(double value1, String value2) throws Throwable {
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void carClassHasCorrectConstructor(double value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, String.class}
+        };
+        Class<?>[] constructorClasses = {double.class, String.class};
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectDoubleValueMessage =
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:87]: change field to attribute

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+                {initialValue, double.class},
+                {"Initial String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setDoubleMethodName, "does not correctly set the value of the", doubleFieldName, "field.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:141]: fix this and all other error messages that have a method name not followed by "method"

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        _assertEquals(initialValue, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value.");
+        _assertEquals("First String", testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value.");
+        Object[][] setDoubleArguments = {
+                {value, double.class}
+        };
+        testClass.callMethod(setDoubleMethodName, setDoubleArguments, classInstance);
+        _assertEquals(value, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value calling the " + setDoubleMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:241]:
```suggestion
                "Your " + getDoubleMethodName + " method does not return the correct value after calling the " + setDoubleMethodName + " method.");
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        Object[][] setBooleanArguments = {
+                {b, String.class}
+        };
+        testClass.callMethod(setStringMethodName, setBooleanArguments, classInstance);
+        _assertEquals(b, testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value calling the " + setStringMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:247]:
```suggestion
                "Your " + getStringMethodName + " method does not return the correct value after calling the " + setStringMethodName + " method.");
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+                "Your " + getDoubleMethodName + " method does not return the correct value.");
+        _assertEquals("First String", testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value.");
+        Object[][] setDoubleArguments = {
+                {value, double.class}
+        };
+        testClass.callMethod(setDoubleMethodName, setDoubleArguments, classInstance);
+        _assertEquals(value, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value calling the " + setDoubleMethodName + " method.");
+        Object[][] setBooleanArguments = {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:242]: this should not be labelled boolean

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q6/QuestionText.md (hunk 2 above)
Comment (SamamthaH) [reply src/oop/special_class_method/hard/q6/QuestionText.md:10]: this still needs changing

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+- Methods:
+    - a `setTopSpeed` method which accepts a `double` value, updates the **topSpeed** attribute with it, and is
+      accessible from other classes
+    - a `getTopSpeed` method which is accessible from other classes and returns the value of the **topSpeed** attribute
+    - a `setType` method which accepts a `String` value, updates the **type** attribute with it, and is accessible
+      from other classes
+    - a `getType` method which is accessible from other classes and returns the value of the **type** attribute
+    - a `toString` method which returns a String of the information about the `Car` class of the form
+      `"This Car has a topSpeed of " + topSpeed " with the type " + type + "."`.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/QuestionText.md:28]: try to follow the x and y substitutions format instead of giving them something they can copy paste in

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+                {b, String.class}
+        };
+        testClass.callMethod(setStringMethodName, setBooleanArguments, classInstance);
+        _assertEquals(b, testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value calling the " + setStringMethodName + " method.");
+        String expectedToString = "This " + className + " has a " + doubleFieldName + " of " + value + " with the " + stringFieldName + " " + b + ".";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, " toString method does not return the correct String after updating the values of its fields using its setter methods.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:260]:
```suggestion
                "Your", className, " toString method does not return the correct String after updating the values of its attributes using its setter methods.");
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+                {initialValue, double.class},
+                {"Initial String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setDoubleMethodName, "does not correctly set the value of the", doubleFieldName, "attribute.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:151]: fix this and all other error messages that have a method name not followed by "method"

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+        assertTrue(testClass.hasModifier(constructorClasses, "public"), incorrectConstructorVisibilityModifier);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void carClassHasCorrectConstructor2(double value1, String value2) throws Throwable {
+        Class<?>[] constructorClasses = {double.class, String.class};
+        Object classInstance = testClass.createInstance();
+        String incorrectDoubleValueMessage =
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " attribute to the correct value.";
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:116]: I feel like since there are two constructors the error message should make it more clear which one is being referred to

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+    @ParameterizedTest
+    @MethodSource("twoStringInputProvider")
+    public void carClassHasCorrectSetTypeMethod(String initialValue, String updatedValue) throws Throwable {
+        Object[][] arguments = {
+                {Math.random() * 1000, double.class},
+                {initialValue, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, initialValue, stringFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:189]: why is this here?

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void carClassHasCorrectGetTypeMethod(double value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, value2, stringFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:209]: why is this here?

**Review by SamamthaH — CHANGES_REQUESTED** [review]
a lot of stuff left unfinished from last review

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+    @Test
+    public void correctTestClass() throws Throwable {
+        Object classInstance = outputClass.createInstance();
+        String[] methodModifiers = {"public"};
+        Object[][] arguments = {
+                {new String[0], String[].class}
+        };
+        outputClass.callMethod("main", arguments, methodModifiers, classInstance, new Clause[]{
+                new StringLiteral("This " + className + " has a " + doubleFieldName + " of " + outputTopSpeed + " with the " + stringFieldName + " " + outputType),
+                new Optional(new StringLiteral(".")),
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:51]: i feel like this shouldn't be optional since it's part of the described toString

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setDoubleMethodName, " method does not correctly set the value of the", doubleFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setDoubleMethodName, setSizeArguments, setMethodModifiers, classInstance);
+        _assertEquals(setValue, testClass.getFieldValue(classInstance, doubleFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setDoubleMethodName, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:153]:
```suggestion
        assertNull(setMethodOutput, String.join(" ", "Your", setDoubleMethodName, "method should not return any output"));
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setStringMethodName, "method does not correctly set the value of the", stringFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setStringMethodName, setMethodArguments, setMethodModifiers, classInstance);
+        _assertEquals(updatedValue, testClass.getFieldValue(classInstance, stringFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setStringMethodName, "should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:196]:
```suggestion
        assertNull(setMethodOutput, String.join(" ", "Your", setStringMethodName, "method should not return any output"));
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void carClassMethodsWorkTogether(double value, String b) throws Throwable {
+        double initialValue = Math.random() * 100;
+        Object[][] arguments = {
+                {initialValue, double.class},
+                {"First String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        _assertEquals(initialValue, testClass.callMethod(getDoubleMethodName, classInstance),
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:239]: no need to check get methods before set methods are called.

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+
+    @BeforeEach
+    public void setUp() {
+        String classString = "oop.special_class_method.hard.q6." + className;
+        testClass = new ObjectTest(classString);
+        String outputClassString = "oop.special_class_method.hard.q6." + testClassName;
+        outputClass = new ObjectTest(outputClassString);
+    }
+
+    @Test
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:42]: Annotate the appropriate tests

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+                new StringLiteral("This " + className + " has a " + doubleFieldName + " of " + outputTopSpeed + " with the " + stringFieldName + " " + outputType + "."),
+                new Optional(new StringLiteral(" ")),
+                new Optional(new NewLine())
+        });
+    }
+
+    @Test
+    public void carClassHasCorrectAttributes() {
+        assertTrue(testClass.hasField(doubleFieldName),
+                "The attribute " + doubleFieldName + " could not be found in your class. Please make sure you have added it, it is spelled correctly.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:59]: the last sentence here and in the other error message does not make grammatical sense

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setDoubleMethodName, "method does not correctly set the value of the", doubleFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setDoubleMethodName, setSizeArguments, setMethodModifiers, classInstance);
+        _assertEquals(setValue, testClass.getFieldValue(classInstance, doubleFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setDoubleMethodName, "method should not return any output"));
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:161]: Switch here and in other places to Void.type with a .hasMethod assert

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+    public void carClassHasCorrectGetTopSpeedMethod(double value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getDoubleMethodName, "method does not correctly get the value of the", doubleFieldName, "attribute.");
+        Object getMethodOutput = testClass.callMethod(getDoubleMethodName, getMethodModifiers, classInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:175]: use .hasMethod first here and in other places

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 26]
```diff
@@ -0,0 +1,36 @@
+# oop-special_class_method-hard-q6
+
+## Question Text
+
+You are writing a Java program that mimics different Vehicles and their attributes. Write a Java class called **Car**
+that has a `double` variable called **topSpeed** and another `String` variable called **type**.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/QuestionText.md:6]: get rid of duplicate class description.

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 27]
```diff
@@ -0,0 +1,345 @@
...
+        String classString = "oop.special_class_method.hard.q6." + className;
+        testClass = new ObjectTest(classString);
+        String outputClassString = "oop.special_class_method.hard.q6." + testClassName;
+        outputClass = new ObjectTest(outputClassString);
+    }
+
+    @Test
+    @Tag("dependent1")
+    @Tag("dependent2")
+    @Tag("dependent3")
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:46]: change this to only one tag "dependency1"

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 28]
```diff
@@ -0,0 +1,345 @@
...
+                Arguments.of(234.4, "Round"),
+                Arguments.of(654.45, "Second"),
+                Arguments.of(43.3, "Fourth")
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    @Tag("dependent1")
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:106]: i don't believe we are tagging constructors, maybe discuss it with seth

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 29]
```diff
@@ -0,0 +1,345 @@
...
+                "Your " + className + " constructor with input does not initialize the " + doubleFieldName + " attribute to the correct value.";
+        String incorrectStringValueMessage =
+                "Your " + className + " constructor with input does not initialize the " + stringFieldName + " attribute to the correct value.";
+        String incorrectConstructorVisibilityModifier =
+                "Your " + className + " constructor does not have the correct visibility modifier.";
+        _assertEquals(value1, testClass.getFieldValue(classInstance, doubleFieldName),
+                incorrectDoubleValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, stringFieldName),
+                incorrectStringValueMessage);
+        assertTrue(testClass.hasModifier(constructorClasses, "public"), incorrectConstructorVisibilityModifier);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:124]: visibility has already been checked in the other test

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 30]
```diff
@@ -0,0 +1,345 @@
...
+    }
+
+    @Test
+    @Tag("dependent2")
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:128]: same here

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 31]
```diff
@@ -0,0 +1,345 @@
...
+                Arguments.of(234.4, 655.3),
+                Arguments.of(654.45, 453.3),
+                Arguments.of(43.3, 654.2)
+
+        );
+    }
+
+    @Test
+    public void setTopSpeedIsDefinedCorrectly() {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q6/MainTest.java:153]: this all already gets checked in the test below, I don't believe there is a need for this test. Same with the rest of the following "isDefinedCorrectly" tests

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
Please address the changes requested by Sam

File: src/oop/special_class_method/hard/q6/QuestionText.md [hunk 32]
```diff
@@ -0,0 +1,31 @@
+# oop-special_class_method-hard-q6
+
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/QuestionText.md:2]: Please add context for the question.

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 33]
```diff
@@ -0,0 +1,339 @@
...
+                "Your " + className + " constructor with parameters does not have the correct parameters.");
+        assertTrue(testClass.hasModifier(classArguments, "public"),
+                "Your " + className + " constructor with parameters does not have the correct modifier.");
+    }
+
+    @Test
+    public void carClassHasRequiredConstructor2() {
+        Class<?>[] classArguments = {};
+        assertTrue(testClass.hasConstructor(classArguments),
+                "Your " + className + " constructor without parameters does not have the correct parameters.");
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:87]: I don't think this error message makes sense. I think it would be better to say something along the lines of "the parameterless constructor could not be found" or "... was not present".

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 34]
```diff
@@ -0,0 +1,339 @@
...
+        assertTrue(testClass.hasModifier(classArguments, "public"),
+                "Your " + className + " constructor without parameters does not have the correct modifier.");
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:89]: Similar thing here, I think we can improve the wording of the error message.

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 35]
```diff
@@ -0,0 +1,339 @@
...
+    public void carClassHasRequiredConstructor() {
+        Class<?>[] classArguments = {double.class, String.class};
+        assertTrue(testClass.hasConstructor(classArguments),
+                "Your " + className + " constructor with parameters does not have the correct parameters.");
+        assertTrue(testClass.hasModifier(classArguments, "public"),
+                "Your " + className + " constructor with parameters does not have the correct modifier.");
+    }
+
+    @Test
+    public void carClassHasRequiredConstructor2() {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:84]:
```suggestion
    public void carClassHasRequiredConstructorWithNoParameters() {
```

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 36]
```diff
@@ -0,0 +1,339 @@
...
+        assertTrue(testClass.hasField(stringFieldName),
+                "The attribute " + stringFieldName + " could not be found in your class. Please make sure you have added it, and it is spelled correctly.");
+        assertTrue(testClass.hasField(stringFieldName, String.class),
+                "Your " + stringFieldName + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(stringFieldName, "private"),
+                "Your " + stringFieldName + " does not have the correct visibility modifier.");
+    }
+
+    @Test
+    public void carClassHasRequiredConstructor() {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:75]:
```suggestion
    public void carClassHasRequiredConstructorWithParameters() {
```

Comment (Hedgemon4) [reply src/oop/special_class_method/hard/q6/MainTest.java:75]: I think this might make it more clear to the student which constructor this is referring to. What do you think?

File: src/oop/special_class_method/hard/q6/MainTest.java (hunk 35 above)
Comment (Hedgemon4) [reply src/oop/special_class_method/hard/q6/MainTest.java:84]: Similar here. Let me know if you think this is an improvement or not.

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 37]
```diff
@@ -0,0 +1,339 @@
...
+                Arguments.of(654.45, "Second"),
+                Arguments.of(43.3, "Fourth")
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    @Tag("dependent1")
+    public void carClassHasCorrectConstructor(double value1, String value2) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:105]: Might want to change the naming here as well.

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 38]
```diff
@@ -0,0 +1,339 @@
...
+                "Your " + className + " constructor with input does not initialize the " + stringFieldName + " attribute to the correct value.";
+        _assertEquals(value1, testClass.getFieldValue(classInstance, doubleFieldName),
+                incorrectDoubleValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, stringFieldName),
+                incorrectStringValueMessage);
+    }
+
+    @Test
+    @Tag("dependent1")
+    public void carClassHasCorrectConstructor2() throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:123]: Also here

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 39]
```diff
@@ -0,0 +1,339 @@
...
+        String expected = "This " + className + " has a " + doubleFieldName + " of " + value + " with the " + stringFieldName + " " + b + ".";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, "toString method does not return the correct String.");
+        Object output = testClass.callMethod("toString", methodModifiers, classInstance);
+        _assertEquals(expected, output, incorrectToStringMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void carClassMethodsWorkTogether(double value, String b) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:310]: We should also maybe add another test to check the setters and getters with the parameterless constructor.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q6/MainTest.java (hunk 36 above)
Comment (KaiKhademi) [reply src/oop/special_class_method/hard/q6/MainTest.java:75]: Yes make sense

File: src/oop/special_class_method/hard/q6/MainTest.java [hunk 40]
```diff
@@ -0,0 +1,366 @@
...
+        String expectedToString = "This " + className + " has a " + doubleFieldName + " of " + value + " with the " + stringFieldName + " " + b + ".";
+        String incorrectToStringMessage = String.join(" ",
+                "Your", className, "toString method does not return the correct String after updating the values of its attributes using its setter methods.");
+        Object output = testClass.callMethod("toString", classInstance);
+        _assertEquals(expectedToString, output, incorrectToStringMessage);
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void carClassWithoutParametersMethodsWorkTogether(double value, String b) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q6/MainTest.java:342]:
```suggestion
    public void carClassMethodsWorkTogether2(double value, String b) throws Throwable {
```

**Review by Hedgemon4 — DISMISSED** [review]
Generally looks good

**Review by Hedgemon4 — APPROVED** [review]
